apps/language_server/test/experimental/source_file_test.exs

```diff
@@ -0,0 +1,775 @@
+               run_changes(orig, [
+                 %{
+                   "text" => "entirely new content",
+                   "range" => invalid_range
+                 }
+               ])
+
+      assert "entirely new content" = text(source)
+    end
+  end
+
+  test "lines" do
+    assert [""] == SourceFile.lines("")
+    assert ["abc"] == SourceFile.lines("abc")
+    assert ["", ""] == SourceFile.lines("\n")
+    assert ["a", ""] == SourceFile.lines("a\n")
+    assert ["", "a"] == SourceFile.lines("\na")
+    assert ["ABCDE", "FGHIJ"] == SourceFile.lines("ABCDE\rFGHIJ")
+    assert ["ABCDE", "FGHIJ"] == SourceFile.lines("ABCDE\r\nFGHIJ")
+    assert ["ABCDE", "", "FGHIJ"] == SourceFile.lines("ABCDE\n\nFGHIJ")
+    assert ["ABCDE", "", "FGHIJ"] == SourceFile.lines("ABCDE\r\rFGHIJ")
+    assert ["ABCDE", "", "FGHIJ"] == SourceFile.lines("ABCDE\n\rFGHIJ")
+  end
+
+  test "full_range" do
+    assert %{
+             "end" => %{"character" => 0, "line" => 0},
+             "start" => %{"character" => 0, "line" => 0}
+           } = SourceFile.full_range(new(""))
+
+    assert %{"end" => %{"character" => 1, "line" => 0}} = SourceFile.full_range(new("a"))
+    assert %{"end" => %{"character" => 0, "line" => 1}} = SourceFile.full_range(new("\n"))
+    assert %{"end" => %{"character" => 2, "line" => 1}} = SourceFile.full_range(new("a\naa"))
+    assert %{"end" => %{"character" => 2, "line" => 1}} = SourceFile.full_range(new("a\r\naa"))
+    assert %{"end" => %{"character" => 8, "line" => 1}} = SourceFile.full_range(new("a\naa🏳️‍🌈"))
+  end
+
+  describe "positions" do
+    test "lsp_position_to_elixir empty" do
+      assert {1, 1} == SourceFile.lsp_position_to_elixir("", {0, 0})
+    end
+
+    test "lsp_position_to_elixir single first char" do
+      assert {1, 1} == SourceFile.lsp_position_to_elixir("abcde", {0, 0})
+    end
+
+    test "lsp_position_to_elixir single line" do
+      assert {1, 2} == SourceFile.lsp_position_to_elixir("abcde", {0, 1})
+    end
+
+    test "lsp_position_to_elixir single line utf8" do
+      assert {1, 2} == SourceFile.lsp_position_to_elixir("🏳️‍🌈abcde", {0, 6})
+    end
+
+    test "lsp_position_to_elixir multi line" do
+      assert {2, 2} == SourceFile.lsp_position_to_elixir("abcde\n1234", {1, 1})
+    end
+
+    test "elixir_position_to_lsp empty" do
+      assert {0, 0} == SourceFile.elixir_position_to_lsp("", {1, 1})
+    end
+
+    test "elixir_position_to_lsp single line first char" do
+      assert {0, 0} == SourceFile.elixir_position_to_lsp("abcde", {1, 1})
+    end
+
+    test "elixir_position_to_lsp single line" do
+      assert {0, 1} == SourceFile.elixir_position_to_lsp("abcde", {1, 2})
+    end
+
+    test "elixir_position_to_lsp single line utf8" do
+      assert {0, 6} == SourceFile.elixir_position_to_lsp("🏳️‍🌈abcde", {1, 2})
+    end
+
+    test "elixir_position_to_lsp multi line" do
+      assert {1, 1} == SourceFile.elixir_position_to_lsp("abcde\n1234", {2, 2})
+    end
+
```

Comment:
Please add this test https://github.com/elixir-lsp/elixir-ls/blob/f02c97a7ba4631d91b0c18f0fd4aac1f8861427f/apps/language_server/test/source_file_test.exs#L665

Reply:
And consider also this https://github.com/elixir-lsp/elixir-ls/blob/f02c97a7ba4631d91b0c18f0fd4aac1f8861427f/apps/language_server/test/source_file_test.exs#L684